### PR TITLE
Implement HashMap iterator next() methods with SIMD

### LINKED ISSUE #13899: Perf improvement idea for HashMap.Iterator.next()
The `Metadata.isUsed` function checks a u1:

https://github.com/ziglang/zig/blob/5238f9c409cdf40db6a1484ab79a7bc518f8f6ad/lib/std/hash_map.zig#L777-L785

`Iterator.next()`:

https://github.com/ziglang/zig/blob/5238f9c409cdf40db6a1484ab79a7bc518f8f6ad/lib/std/hash_map.zig#L823-L843

We could load the `Metadata` as an `@Vector(u8, 16)` and then find the first index of the element with the final bit set

### PR DESCRIPTION
**Pull request by candrewlee14**
Should resolve #13899.
Here is a benchmark on an example that inserts a million entries, deletes most of the entries, then iterates over all the entries to sum them up. This is a modest perf improvement for remove-heavy workloads. This specific example runs about 10-15% faster on my x86 machine.

image

Compilation commands:
```bash
zig build-exe src/example2.zig --zig-lib-dir ../zig/lib -femit-bin=example2-new -O ReleaseSafe
``` 
```bash
zig build-exe src/example2.zig -femit-bin=example2-old -O ReleaseSafe
```

The benchmarked code example: 
```zig
const std = @import("std");

pub fn main() !void {
    const Point = struct { x: i32, y: i32 };
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    var map = std.AutoHashMap(u32, Point).init(
        gpa.allocator(),
    );
    defer map.deinit();

    var prng = std.rand.DefaultPrng.init(0);
    const rand = prng.random();

    const NUM_INSERTS: usize = 1_000_000;
    // insert lots of entries
    {
        var i: usize = 0;
        while (i < NUM_INSERTS) : (i += 1) {
            try map.put(rand.int(u32), .{ .x = @rem(rand.int(i32), 100), .y = @rem(rand.int(i32), 100) });
        }
    }
    // delete 99/100 entries
    {
        var iterator = map.iterator();
        var i: usize = 0;
        while (iterator.next()) |*entry| : (i += 1) {
            if (i % 100 != 0) {
                map.removeByPtr(entry.key_ptr);
            }
        }
    }
    // sum up
    var count: usize = 0;
    var sum = Point{ .x = 0, .y = 0 };
    {
        var iterator = map.iterator();
        while (iterator.next()) |entry| : (count += 1) {
            sum.x += entry.value_ptr.x;
            sum.y += entry.value_ptr.y;
        }
    }
    std.debug.print("sum: {}, entries: {}\n", .{ sum, count });
}

```

### CODE REVIEW

**Review by jacobly0 — CHANGES_REQUESTED** [review]
I believe this should work around the CI failures.  If this PR gets merged with these changes, make sure a comment is added to #13782 mentioning the new references.  I think adding `#13782` to the commit message should do this automatically.

File: lib/std/hash_map.zig [hunk 1]
```diff
@@ -818,26 +818,54 @@ pub fn HashMapUnmanaged(
...
-                    it.index += 1;
-                }) {
-                    if (metadata[0].isUsed()) {
-                        const key = &it.hm.keys()[it.index];
-                        const value = &it.hm.values()[it.index];
+                if (comptime builtin.zig_backend != .stage2_c and // TODO
+                    // https://github.com/ziglang/zig/issues/13782
+                    (builtin.zig_backend != .stage2_llvm or builtin.cpu.arch.endian() != .Big))
+                {
+                    comptime var vec_size: u8 = 16;
```
Comment (IntegratedQuantum) [inline lib/std/hash_map.zig:835]: I think this should probably be using `std.simd.suggestVectorSize` to optimize further when for example AVX or newer is targeted.

**Review by jacobly0 — COMMENTED** [review]
I benchmarked the following two possible implementations:

Method 1: `ctz(input < 0)`
This is essentially what this PR's code optimizes to.

Method 2: `min(select(input >= 0, ~0, 0) | iota)`
This was just an educated guess on my part.

On x86_64, we get lucky with Method 1 because we happen to have two instructions that compute these exact operations.
On aarch64, while we do have instructions for the two operations in Method 1 individually, the bitCast between a bool vector and an integer explodes into 47 (!) instructions.

On x86_64, while we don't have reduce instructions for Method 2 in baseline, the instruction bloat is O(log(n)) rather than O(n) and comes in at a much more reasonable 10 instructions, and by skylake this seems to be reduced even further to 2 instructions.
On aarch64, Method 2 is great because it exactly corresponds with 3 instructions.

| | Method 1 | Method 2 |
|-|-|-|
| x86_64 baseline | ~1ns | ~1.5ns |
| x86_64 native | ~0.75ns | ~0.75ns |
| aarch64 | ~6.66ns | ~1ns |

I'm not sure how to best handle x86, but leaving a 6x speedup on the table for aarch64 doesn't seem great.

[Here](https://godbolt.org/z/b5rfdddj7) is the specific codegen I am referencing since my expressions above are probably somewhat ambiguous.

**Comment by candrewlee14** [comment]
@jacobly0 Should I do a comptime switch on the arch then and use a different method for x86 and arm?

**Comment by jacobly0** [comment]
> @jacobly0 Should I do a comptime switch on the arch then and use a different method for x86 and arm?

My guess is that Method 1 will only ever be beneficial on older x86_64, so it seems like the best options are to either use Method 2 and lose a .5x speedup on old cpus, or use Method 1 for x86/x86_64 and Method 2 for everything else.

**Comment by candrewlee14** [comment]
image
Here's the results of each implementation method for my x86_64 machine. I think I'll do a `comptime if` on arch to get the best of both worlds on that.

File: lib/std/hash_map.zig [hunk 2]
```diff
@@ -852,14 +885,50 @@ pub fn HashMapUnmanaged(
...
-                        const item = &self.items[0];
-                        self.metadata += 1;
-                        self.items += 1;
-                        if (used) {
-                            return item;
+                    if (comptime builtin.zig_backend != .stage2_c and // TODO
+                        // https://github.com/ziglang/zig/issues/13782
+                        (builtin.zig_backend != .stage2_llvm or builtin.cpu.arch.endian() != .Big))
+                    {
+                        comptime var vec_size: usize = std.simd.suggestVectorSize(u16) orelse 1;
```
Comment (jacobly0) [inline lib/std/hash_map.zig:892]:
```suggestion
                        comptime var vec_size: usize = std.simd.suggestVectorSize(u8) orelse 1;
```

**Review by jacobly0 — COMMENTED** [review]
Here are my results for the benchmark program:
| | ctz | reduce | unvectorized | ctz unvectorized |
|-|-|-|-|-|
x86_64 native | 744 ms | 755 ms | 753 ms | 725 ms |
x86_64 baseline | 747 ms | 759 ms | 771 ms | 742 ms |
aarch64 | 1.2 s | 1.11 s | 1.03 s | 1.06s |

Where the first three methods are in the same order as in the code and the last one is the one I mentioned earlier.

**Comment by IntegratedQuantum** [comment]
I would like to step in here and criticize your benchmarking methodology.
Some of the performance differences mentioned in this thread are below 5%.
Differences on that scale can also be caused by differences in code alignment. (See ["Performance Matters" by Emery Berger](https://www.youtube.com/watch?v=r-TLSBdHe1A))

A better approach would be to only measure the part we actually care about instead of also measuring all the initialization code.
This can be done using for example `std.time.nanoTimestamp()`:
```zig
pub fn main() !void {
    ... // Initialization code
    // sum up
    var t1 = std.time.nanoTimestamp();
    var count: usize = 0;
    var sum = Point{ .x = 0, .y = 0 };
    {
        var iterator = map.iterator();
        while (iterator.next()) |entry| : (count += 1) {
            sum.x += entry.value_ptr.x;
            sum.y += entry.value_ptr.y;
        }
    }
    var t2 = std.time.nanoTimestamp();
    std.debug.print("sum: {}, entries: {}, time: {d:.2} ms\n", .{ sum, count, @intToFloat(f32, t2 - t1)/1000000.0 });
}
```
Performance results (averaged over 4 iterations):
| | ctz | reduce | unvectorized| bonus: unvectorized + unrelated code changes |
|-|-|-|-|-|
| x86_64 native | 0.64 ms | 0.97 ms | 1.79 ms | 1.75 ms |
| x86_64 baseline| 1.21 ms | 1.30 ms | 1.75 ms | 1.80 ms |

The results give a much clearer image and the relative differences are more significant.

To give you a rough idea of how big of a deal code alignment is I made some unrelated changes to the initialization process and measured again (last column). As you can see the results changed by 2-3% in this case, although neither the measured code nor the data has changed.

**Comment by andrewrk** [comment]
Thank you for doing this work.

I think this complexity is not worth it, given that it makes more sense to use an ArrayHashMap when iteration is desired. What I mean by complexity: the CI is failing, the implementation contains many branches and edge cases, and the existing unit test coverage is insufficient to be confident that this code is bug free.

I am willing to reconsider this if someone can show that using this patch in a real world project provides better results than switching to ArrayHashMap.

**Pull request closed without merging**